lib/Sema/CSWalker.cpp

```diff
@@ -205,6 +205,67 @@ class LocalCSWalker : public glu::ast::ASTWalker<LocalCSWalker, void> {
             )
         );
     }
+
+    void postVisitUnaryOpExpr(glu::ast::UnaryOpExpr *node)
+    {
+        auto *operandTy = node->getOperand()->getType();
+        auto *resultTy = node->getType();
+
+        auto &arena = node->getModule()->getContext()->getTypesMemoryArena();
+
+        auto *expectedFunctionTy = arena.create<glu::types::FunctionTy>(
+            llvm::ArrayRef<glu::types::TypeBase *>(&operandTy, 1), resultTy
+        );
+
+        llvm::StringRef opName = node->getOperator().getLexeme();
+        auto *item = _cs.getScopeTable()->lookupItem(opName);
+        llvm::ArrayRef<ast::DeclBase *> overloadDecls = item
+            ? llvm::makeArrayRef(item->decls)
+            : llvm::ArrayRef<ast::DeclBase *>();
+
+        llvm::SmallVector<Constraint *, 4> constraints;
+
+        for (auto *decl : overloadDecls) {
+            if (auto *fnDecl = llvm::dyn_cast<ast::FunctionDecl>(decl)) {
+
+                if (fnDecl->getParamCount() != 1)
+                    continue;
```

Comment:
that works because we only consider function declarations with exactly one argument. but this doesn't use the constraint system, which means that the constraint resolution system doesn't know about the non-matches, and error messages will not have the whole picture.
To get better error messages, leave out those conditions, and let the solver handle it.